Improved C++ command line interface

alranel opened the pull request:
This implements the improved command line interface as discussed in #4591 and slic3r/Slic3r-Manual#60

lordofhyphens commented on xs/src/libslic3r/TriangleMesh.cpp:
```diff
@@ -438,7 +434,7 @@ Pointf3s TriangleMesh::normals() const
     if (this->repaired) {
         for (auto i = 0; i < stl.stats.number_of_facets; i++) {
-            const auto& n {stl.facet_start[i].normal};
+            const auto& n = stl.facet_start[i].normal;
```
Inconsistency with TriangleMesh.cpp:407, 423, etc.

alranel replied:
Fixed.

lordofhyphens commented on xs/src/libslic3r/SimplePrint.hpp:
```diff
@@ -0,0 +1,28 @@
+
+namespace Slic3r {
+
+class SimplePrint {
```
I think users of libslic3r would appreciate a doxygen comment explaining what is provided/expected with this class.

lordofhyphens commented on xs/src/libslic3r/SimplePrint.hpp:
```diff
@@ -0,0 +1,28 @@
+class SimplePrint {
+    public:
+    bool center{true};
+    std::function<void(int, const std::string&)> status_cb {nullptr};
```
I think this can be tested similar to how I test some of the other callback methods via a lambda that sets a value to true/false as expected.

lordofhyphens commented on xs/src/libslic3r/SimplePrint.hpp:
```diff
@@ -0,0 +1,28 @@
+    double total_used_filament() const { return this->_print.total_used_filament; }
+    double total_extruded_volume() const { return this->_print.total_extruded_volume; }
+    void set_model(const Model &model);
+    void export_gcode(std::string outfile);
```
I'm wondering if a regression test is feasible here, given a fixed initial config. Could we split this method so it outputs a string and tests some of the CLI-only parameters via GCodeReader? I think that would be resilient enough from changes elsewhere in Slic3r.

lordofhyphens commented on xs/src/libslic3r/SimplePrint.hpp:
```diff
@@ -0,0 +1,28 @@
+    void export_gcode(std::string outfile);
+    
+    private:
+    Model _model;
```
For testability, I'd really like to be able to reach inside of _model and _print. Can we safely expose both via const reference methods? It is probably cleaner than adding `friend`s of the test system.

lordofhyphens commented:
For tests, I think there should be some basic checks that each of the switches documented in 
https://github.com/slic3r/Slic3r-Manual/blob/382ceed58407b17c5b7653bb0534900ae54498ce/src/advanced/command-line.md 

do what they say they do, except for probably datadir and autosave. 

It is sufficient to verify that a file gets created, I think, for each of the --export methods.

Most of the transforms can be tested via sla_svg export, which should be a lot more stable in its output than gcode output.

alranel commented on src/slic3r.cpp:
```diff
@@ -110,151 +108,308 @@ main(int argc, char **argv)
+    if (actions.empty()) {
+#ifdef USE_WX
+        GUI::App *gui = new GUI::App();
+        gui->autosave = this->config.getString("autosave");
```
Fixed.

alranel commented on xs/src/libslic3r/Config.cpp:
```diff
@@ -22,41 +21,36 @@ Config::new_from_defaults(std::initializer_list<std::string> init)
 Config::new_from_defaults(t_config_option_keys init)
 {
-    auto my_config(std::make_shared<Config>());
+    std::shared_ptr<Config>  my_config(std::make_shared<Config>());
```
I can't remember the exact detail, but whenever I changed from `auto` to the explicit type, except when clearly useful for clarity, was for making gcc 4.9 happy.

alranel commented on xs/src/libslic3r/Config.cpp:
```diff
@@ -91,28 +85,28 @@ void
 {
     try {
-        const auto& def {print_config_def.options.at(opt_key)};
+        const auto& def = print_config_def.options.at(opt_key);
```
`auto foo = { xx }` does not work on gcc 4.9 because the type of `foo` is not the right type but always `std::initializer_list<T>`, so the first Travis build fails. I think this is because gcc 4.9 supports C++14 but not C++17 which made this smarter.
This PR removes the `#ifndef SLIC3RXS` guards, so some C++17 code is now exposed to gcc 4.9 on Travis. I felt more comfortable with changing many `auto foo = { xx }` lines than changing the Travis build chain which has cached dependencies and other joys.

alranel commented on src/slic3r.hpp:
```diff
@@ -0,0 +1,29 @@
+    std::vector<Model> models;
+    
+    void print_help(bool include_print_options = false) const;
+    void export_models(IO::ExportFormat format);
```
Done.

alranel commented on xs/src/libslic3r/PrintConfig.hpp:
```diff
@@ -639,78 +639,28 @@ class SLAPrintConfig
 };
 
-class CLIConfigDef : public ConfigDef
+class CLIActionsConfigDef : public ConfigDef
```
Done.

alranel commented on xs/src/libslic3r/PrintGCode.cpp:
```diff
@@ -25,70 +18,70 @@ PrintGCode::output()
 
     // Writes notes (content of all Settings tabs -> Notes)
-    fh << gcodegen.notes();
+    fh << _gcodegen.notes();
```
I totally understand, but I had the `auto foo = { x }` issue mentioned above and this looked like an easy way to do it. I'm not a fan of implied `->` but I think it's fine for underscore variables like `_gcodegen`.

alranel commented:
I did most things you suggested. Regarding tests, I totally agree with you. I just have no time now for working on tests for this. Since this is a large PR I suggest to merge it as it is and work on test with further PRs.  I'd like to move onto other tasks which I find more urgent (QA of the C++ port, removing duplicated Perl/C++ code...).

lordofhyphens commented:
`--output-filename-format FORMAT` is not implemented. 
I've added the most basic of tests for the export functions and pushed to your branch. 

The SVG output test is probably wrong in that it probably (?) should be exporting a bunch of file slices.

lordofhyphens commented:
I've pushed one more update to set the SVG export tests to behave as per the manual's entry.

lordofhyphens commented:
I'll have one more test update for you for --dont-arrange and --center; I noticed that an exported config from the current (perl) GUI is causing UnknownOptionExceptions on load.

alranel commented:
* `--output-filename-format FORMAT` is actually implemented and seems to work.
* I added the missing definition for `--export-amf`.
* I fixed `--center X,Y`, `--dont-arrange` and other things.
* What's the issue with SVG?

I wasn't able to run your new tests.

lordofhyphens commented:
SVG, according to the manual page, says it puts the layer number in the filename by default

alranel commented:
I fixed the documentation regarding `--export-sla-svg`; the implementation is correct.
